In [ ]:
from google.colab import drive
drive.mount ('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv('/content/drive/MyDrive/EDA2026-3/praktikum-1 dikelas/penjualan_umkm_2024.csv')
df.head()

,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar
0,TRX2024198,2024-11-14,Jakarta Selatan,Makanan,Sambal Roa,8,38000,5,Transfer,4.7,288800
1,TRX2025061,2024-08-16,Depok,Fashion,Kaos Lokal,8,131500,15,Tunai,3.8,894200
2,TRX2024249,2024-12-21,Depok,Fashion,Hijab Motif,12,102500,10,E-Wallet,4.3,1107000
3,TRX2024903,2024-09-27,Bekasi,Fashion,Kemeja Tenun,11,113500,10,E-Wallet,NaN,1123650
4,TRX2024097,2024-05-06,Depok,Kerajinan,Batik Tulis,8,123000,20,QRIS,4.5,787200


> **TODO A :**

In [ ]:
# a. Tampilkan 5 baris TERAKHIR dari df
df.tail()

,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar
1207,TRX2025021,2024-11-18,Tangerang Selatan,Makanan,Keripik Singkong,1,22000,20,QRIS,4.1,17600
1208,TRX2024789,2024-09-16,Bogor,Minuman,Teh Herbal,5,55000,10,Transfer,NaN,247500
1209,TRX2024256,2024-08-13,Bogor,Minuman,Sirup Markisa,2,47000,10,QRIS,3.6,84600
1210,TRX2024968,2024-01-08,Bekasi,Kerajinan,Tas Anyaman,7,132000,20,Transfer,4.6,739200
1211,TRX2024952,2024-03-11,Bekasi,Minuman,Teh Herbal,7,29500,20,Transfer,4.4,165200


In [ ]:
# b. Tampilkan daftar nama kolom beserta tipe datanya
df.columns

Index(['id_transaksi', 'tanggal', 'kota', 'kategori', 'produk', 'jumlah',
       'harga_satuan', 'diskon_persen', 'metode_bayar', 'rating_pelanggan',
       'total_bayar'],
      dtype='object')

In [ ]:
# Expected: 1212 baris x 11 kolom; kolom 'tanggal' bertipe object (belum datetime)
df.shape

(1212, 11)

> **TODO B :**

In [ ]:
# a. Hitung jumlah nilai kosong (missing) per kolom
print("Jumlah nilai kosong per kolom:")
print(df.isnull().sum())

Jumlah nilai kosong per kolom:
id_transaksi         0
tanggal              0
kota                 0
kategori             0
produk               0
jumlah               0
harga_satuan         0
diskon_persen        0
metode_bayar        25
rating_pelanggan    61
total_bayar          0
dtype: int64


In [ ]:
# b. Hitung jumlah baris duplikat
df.duplicated().sum()

np.int64(12)

In [ ]:
# c. Hitung berapa nilai unik pada kolom 'kota'
df['kota'].unique()

array(['Jakarta Selatan', 'Depok', 'Bekasi', 'Tangerang Selatan', 'DEPOK',
       'Bogor', 'BOGOR', 'JAKARTA SELATAN', 'TANGERANG SELATAN', 'BEKASI'],
      dtype=object)

---
## Bagian B — Seleksi dan filter (Latihan 2, berpasangan)
**Tujuan:** mengambil bagian data yang relevan dengan pertanyaan analisis.

In [ ]:
# DEMO B1 — memilih kolom
df['total_bayar'].head()
df[['kota','kategori','total_bayar']].head()

,kota,kategori,total_bayar
0,Jakarta Selatan,Makanan,288800
1,Depok,Fashion,894200
2,Depok,Fashion,1107000
3,Bekasi,Fashion,1123650
4,Depok,Kerajinan,787200


In [ ]:
# DEMO B2 — loc (berbasis label) vs iloc (berbasis posisi)
print(df.loc[0, 'produk'])
print(df.iloc[0, 4])
df.loc[0:4, ['produk','jumlah','total_bayar']]

Sambal Roa
Sambal Roa


,produk,jumlah,total_bayar
0,Sambal Roa,8,288800
1,Kaos Lokal,8,894200
2,Hijab Motif,12,1107000
3,Kemeja Tenun,11,1123650
4,Batik Tulis,8,787200


In [ ]:
# DEMO B3 — filter dengan kondisi (boolean mask)
mask = df['total_bayar'] > 1_000_000
print('Transaksi di atas 1 juta:', mask.sum())
df[mask].head()

Transaksi di atas 1 juta: 180


,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar
2,TRX2024249,2024-12-21,Depok,Fashion,Hijab Motif,12,102500,10,E-Wallet,4.3,1107000
3,TRX2024903,2024-09-27,Bekasi,Fashion,Kemeja Tenun,11,113500,10,E-Wallet,NaN,1123650
36,TRX2024121,2024-05-31,Bekasi,Kerajinan,Batik Tulis,13,132000,5,Tunai,4.4,1630200
39,TRX2024692,2024-08-07,Depok,Kerajinan,Gerabah Hias,10,107000,0,Transfer,4.4,1070000
42,TRX2024232,2024-08-26,Tangerang Selatan,Kerajinan,Tas Anyaman,14,135500,0,QRIS,3.6,1897000


In [ ]:
# DEMO B4 — menggabungkan kondisi
hi = df[(df['kategori'] == 'Makanan') & (df['diskon_persen'] >= 10)]
print(hi.shape)
df[df['metode_bayar'].isin(['QRIS','E-Wallet'])].shape

(201, 11)


(648, 11)

> **TODO B-5

In [ ]:
# a. Ambil semua transaksi dengan rating_pelanggan < 3
rating_rendah = df[df['rating_pelanggan'] < 3]

print("Transaksi dengan rating pelanggan < 3:")
display(rating_rendah)

Transaksi dengan rating pelanggan < 3:


,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar
25,TRX2025090,2024-11-27,Tangerang Selatan,Minuman,Sirup Markisa,6,55500,10,QRIS,2.8,299700
148,TRX2024429,2024-06-08,Jakarta Selatan,Fashion,Kaos Lokal,2,100500,0,Tunai,2.9,201000
231,TRX2024955,2024-01-30,Bogor,Makanan,Sambal Roa,12,26000,10,E-Wallet,2.5,280800
244,TRX2024349,2024-10-11,Depok,Minuman,Sirup Markisa,8,44500,0,Transfer,2.8,356000
361,TRX2024423,2024-09-05,Bogor,Minuman,Teh Herbal,12,46000,10,E-Wallet,2.9,496800
368,TRX2024214,2024-05-07,Depok,Minuman,Teh Herbal,4,48500,0,Tunai,2.8,194000
383,TRX2024188,2024-01-22,Depok,Fashion,Hijab Motif,6,118500,0,Tunai,2.9,711000
455,TRX2024886,2024-03-24,Jakarta Selatan,Makanan,Kue Kering,7,26000,5,E-Wallet,2.9,172900
471,TRX2024229,2024-11-03,Depok,Kerajinan,Batik Tulis,6,176000,0,Transfer,2.9,1056000
477,TRX2024751,2024-07-26,Tangerang Selatan,Minuman,Sirup Markisa,11,32500,15,QRIS,2.9,303875


In [ ]:
# b. Berapa banyak transaksi tersebut dan berapa rata-rata total_bayar-nya?
print("Jumlah transaksi:", len(rating_rendah))
print("Rata-rata total_bayar:", rating_rendah['total_bayar'].mean())

Jumlah transaksi: 19
Rata-rata total_bayar: 386563.15789473685


In [ ]:
# c. Ambil transaksi kategori 'Kerajinan' ATAU 'Fashion' dengan jumlah >= 10
transaksi_kategori = df[
    (df['kategori'].isin(['Kerajinan', 'Fashion'])) &
    (df['jumlah'] >= 10)
]

print("\nTransaksi kategori Kerajinan atau Fashion dengan jumlah >= 10:")
display(transaksi_kategori)


Transaksi kategori Kerajinan atau Fashion dengan jumlah >= 10:


,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar
2,TRX2024249,2024-12-21,Depok,Fashion,Hijab Motif,12,102500,10,E-Wallet,4.3,1107000
3,TRX2024903,2024-09-27,Bekasi,Fashion,Kemeja Tenun,11,113500,10,E-Wallet,NaN,1123650
36,TRX2024121,2024-05-31,Bekasi,Kerajinan,Batik Tulis,13,132000,5,Tunai,4.4,1630200
39,TRX2024692,2024-08-07,Depok,Kerajinan,Gerabah Hias,10,107000,0,Transfer,4.4,1070000
42,TRX2024232,2024-08-26,Tangerang Selatan,Kerajinan,Tas Anyaman,14,135500,0,QRIS,3.6,1897000
...,...,...,...,...,...,...,...,...,...,...,...
1132,TRX2024076,2024-03-14,Tangerang Selatan,Kerajinan,Gerabah Hias,14,149000,15,QRIS,5.0,1773100
1138,TRX2024503,2024-09-17,Depok,Kerajinan,Tas Anyaman,12,117500,10,QRIS,4.1,1269000
1152,TRX2024036,2024-11-18,Bogor,Kerajinan,Gerabah Hias,12,106000,0,QRIS,5.0,1272000
1177,TRX2024704,2024-07-31,Bekasi,Kerajinan,Gerabah Hias,10,162000,0,QRIS,3.5,1620000


> **TODO B-6 : Pertanyaan Bisnis**


In [ ]:
transaksi_tunai_besar = df[
    (df['metode_bayar'] == 'Tunai') &
    (df['total_bayar'] > 500_000)
]

transaksi_tunai_besar

,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar
1,TRX2025061,2024-08-16,Depok,Fashion,Kaos Lokal,8,131500,15,Tunai,3.8,894200
22,TRX2024255,2024-04-01,Bogor,Fashion,Kaos Lokal,5,119000,0,Tunai,5.0,595000
35,TRX2024509,2024-02-26,Bekasi,Kerajinan,Batik Tulis,4,167500,0,Tunai,NaN,670000
36,TRX2024121,2024-05-31,Bekasi,Kerajinan,Batik Tulis,13,132000,5,Tunai,4.4,1630200
44,TRX2025006,2024-08-11,Bogor,Kerajinan,Tas Anyaman,10,116500,0,Tunai,3.6,1165000
...,...,...,...,...,...,...,...,...,...,...,...
1121,TRX2024483,2024-09-24,Bogor,Kerajinan,Gerabah Hias,13,103000,0,Tunai,4.7,1339000
1136,TRX2024728,2024-12-13,Jakarta Selatan,Fashion,Kaos Lokal,9,134000,0,Tunai,4.0,1206000
1156,TRX2024497,2024-07-28,Bogor,Fashion,Kemeja Tenun,9,85500,15,Tunai,4.3,654075
1168,TRX2024887,2024-02-05,Jakarta Selatan,Kerajinan,Gerabah Hias,4,129500,0,Tunai,NaN,518000


In [ ]:
hasil_b6 = transaksi_tunai_besar[
    ['kota', 'produk', 'jumlah', 'total_bayar']
].sort_values(
    'total_bayar',
    ascending=False
).head(10)

hasil_b6

,kota,produk,jumlah,total_bayar
385,Bekasi,Batik Tulis,14,2667000
97,Bekasi,Gerabah Hias,12,2286000
742,Jakarta Selatan,Batik Tulis,13,2242500
988,Bogor,Batik Tulis,14,2237200
398,Jakarta Selatan,Kemeja Tenun,14,2093000
129,Bogor,Batik Tulis,13,2054000
116,Bekasi,Gerabah Hias,14,1999200
252,Depok,Hijab Motif,14,1988000
168,Bogor,Gerabah Hias,13,1982500
1000,Jakarta Selatan,Gerabah Hias,11,1947000


---
## Bagian C — Kolom turunan dan pengurutan

In [ ]:
# DEMO C1 — membuat kolom baru
df['harga_bruto'] = df['jumlah'] * df['harga_satuan']
df['hemat'] = df['harga_bruto'] - df['total_bayar']
df[['harga_bruto','total_bayar','hemat']].head()

,harga_bruto,total_bayar,hemat
0,304000,288800,15200
1,1052000,894200,157800
2,1230000,1107000,123000
3,1248500,1123650,124850
4,984000,787200,196800


In [ ]:
# DEMO C2 — kolom tanggal
df['tanggal'] = pd.to_datetime(df['tanggal'])
df['bulan'] = df['tanggal'].dt.month
df['nama_hari'] = df['tanggal'].dt.day_name()
df[['tanggal','bulan','nama_hari']].head()

,tanggal,bulan,nama_hari
0,2024-11-14,11,Thursday
1,2024-08-16,8,Friday
2,2024-12-21,12,Saturday
3,2024-09-27,9,Friday
4,2024-05-06,5,Monday


In [ ]:
print(df.columns.tolist())

['id_transaksi', 'tanggal', 'kota', 'kategori', 'produk', 'jumlah', 'harga_satuan', 'diskon_persen', 'metode_bayar', 'rating_pelanggan', 'total_bayar', 'kelas_nilai']


In [ ]:
# a. kolom 'kelas_nilai'
df['kelas_nilai'] = np.where(
    df['total_bayar'] > 500_000,
    'Besar',
    'Kecil'
)

In [ ]:
# b. Hitung berapa banyak transaksi pada tiap kelas
print("Jumlah transaksi tiap kelas:")
print(df['kelas_nilai'].value_counts())

Jumlah transaksi tiap kelas:
kelas_nilai
Kecil    845
Besar    367
Name: count, dtype: int64


In [ ]:
# c. Hitung penghematan setiap transaksi
df['hemat'] = (
    df['harga_satuan'] * df['jumlah']
) * df['diskon_persen'] / 100


In [ ]:
print("\n5 transaksi dengan penghematan terbesar:")

display(
    df.sort_values(
        'hemat',
        ascending=False
    ).head(5)
)


5 transaksi dengan penghematan terbesar:


,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar,kelas_nilai,hemat
295,TRX2024224,2024-02-13,Jakarta Selatan,Kerajinan,Gerabah Hias,14,179500,20,Transfer,4.1,2010400,Besar,502600.0
155,TRX2024618,2024-06-15,Jakarta Selatan,Kerajinan,Batik Tulis,14,169500,20,E-Wallet,3.8,1898400,Besar,474600.0
220,TRX2024902,2024-02-04,Jakarta Selatan,Kerajinan,Batik Tulis,13,182500,20,QRIS,4.1,1898000,Besar,474500.0
112,TRX2025135,2024-07-05,Tangerang Selatan,Kerajinan,Gerabah Hias,13,176500,20,Transfer,4.0,1835600,Besar,458900.0
381,TRX2025146,2024-10-08,Bogor,Kerajinan,Gerabah Hias,13,163000,20,Transfer,NaN,1695200,Besar,423800.0


---
## Bagian D — Agregasi dengan groupby
Pola yang akan Anda pakai sepanjang semester: **split → apply → combine**.

In [ ]:
# DEMO D1 — agregasi sederhana
df.groupby('kategori')['total_bayar'].mean().round(0)

In [ ]:
# DEMO D2 — beberapa agregasi sekaligus
ringkas = (df.groupby('kategori')
             .agg(jumlah_transaksi=('id_transaksi','count'),
                  total_penjualan=('total_bayar','sum'),
                  rata_rata_nilai=('total_bayar','mean'),
                  rata_rata_rating=('rating_pelanggan','mean'))
             .round(2)
             .sort_values('total_penjualan', ascending=False))
ringkas

In [ ]:
# DEMO D3 — dua kunci pengelompokan
df.groupby(['kategori','metode_bayar'])['total_bayar'].sum().unstack(fill_value=0)

In [ ]:
# a. Total penjualan per kota, diurutkan dari yang terbesar
total_penjualan_kota = (
    df.groupby('kota')['total_bayar']
      .sum()
      .sort_values(ascending=False)
)

print("a. Total penjualan per kota:")
print(total_penjualan_kota)

a. Total penjualan per kota:
kota
Depok                172719800
Jakarta Selatan      133467675
Bogor                105871075
Bekasi                95609225
Tangerang Selatan     65794250
JAKARTA SELATAN        5236300
BEKASI                 4139000
DEPOK                  3962400
BOGOR                  1136250
TANGERANG SELATAN      1091750
Name: total_bayar, dtype: int64


In [ ]:
# b. Rata-rata rating per kategori
rata_rating_kategori = (
    df.groupby('kategori')['rating_pelanggan']
      .mean()
      .sort_values()
)

print("\nb. Rata-rata rating per kategori:")
print(rata_rating_kategori)

print(
    "\nKategori dengan rating paling rendah:",
    rata_rating_kategori.idxmin()
)



b. Rata-rata rating per kategori:
kategori
Fashion      4.122162
Minuman      4.125455
Kerajinan    4.142553
Makanan      4.185965
Name: rating_pelanggan, dtype: float64

Kategori dengan rating paling rendah: Fashion


In [ ]:
# c. Penjualan per bulan
penjualan_bulan = (
    df.groupby('bulan')['total_bayar']
      .sum()
      .sort_values(ascending=False)
)

print("\nc. Penjualan per bulan:")
print(penjualan_bulan)

print(
    "\nBulan dengan penjualan tertinggi:",
    penjualan_bulan.idxmax()
)

In [ ]:
# TODO D5 — satu temuan
# Tulis SATU kalimat temuan dari hasil di atas dalam sel Markdown berikutnya.
# Format: 'X memiliki ... sebesar ..., lebih tinggi/rendah dibandingkan ...'
# Ingat: ini temuan DESKRIPTIF, belum boleh diklaim sebagai sebab-akibat.


In [ ]:
# TODO D5 — satu temuan

# Temuan berdasarkan total penjualan per kota
total_penjualan_kota = (
    df.groupby('kota')['total_bayar']
      .sum()
      .sort_values(ascending=False)
)

kota_tertinggi = total_penjualan_kota.index[0]
nilai_tertinggi = total_penjualan_kota.iloc[0]

kota_terendah = total_penjualan_kota.index[-1]
nilai_terendah = total_penjualan_kota.iloc[-1]

print(
    f"{kota_tertinggi} memiliki total penjualan sebesar "
    f"{nilai_tertinggi:,.0f}, lebih tinggi dibandingkan "
    f"{kota_terendah} sebesar {nilai_terendah:,.0f}."
)

Depok memiliki total penjualan sebesar 172,719,800, lebih tinggi dibandingkan TANGERANG SELATAN sebesar 1,091,750.


**Temuan saya:** _(tulis di sini)_


Depok memiliki total penjualan sebesar 172,719,800, lebih tinggi dibandingkan TANGERANG SELATAN sebesar 1,091,750.
